client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,789 @@
+   * Another common example is in create views, where the form is likely invalid
+   * on initial load due to missing required fields (e.g. `title`).
+   *
+   * When this happens, Wagtail will not update the preview data in the session,
+   * which means it may still contain the outdated preview data from the
+   * previous full-page load. We want to clear this data immediately so that the
+   * preview panel displays the "Preview is not available" screen instead of an
+   * outdated preview.
+   *
+   * This flag determines whether the preview is "ready" for further updates –
+   * i.e. this is true if the preview data has been cleared after an invalid
+   * initial load, or if the preview data is already valid on initial load.
+   *
+   * An alternative approach would be to handle the initial state of the
+   * session's preview data in the backend, but this would require the logic to
+   * be applied in all the different places (i.e. page and snippets create and
+   * edit views).
+   */
+  ready = false;
+  /** Whether the preview is currently available. This is used to distinguish
+   * whether we are rendering a preview or the "Preview is not available"
+   * screen. So even if the preview is currently outdated, this is still `true`
+   * as long as the preview data is available and the preview is rendered (e.g.
+   * if the form becomes invalid after the preview is successfully rendered).
+   */
+  available = true;
+  /** Promise for the current update request. This is resolved as soon as the
+   * update request is successful, so the preview iframe may not have been
+   * fully reloaded.
+   */
+  updatePromise: Promise<boolean> | null = null;
+  /** Serialized form payload to be compared in between intervals to determine
+   * whether an update should be performed. Note that we currently do not handle
+   * file inputs.
+   */
+  formPayload = '';
+
+  /**
+   * The default size input element.
+   * This is the size input element with the `data-default-size` data attribute.
+   * If no input element has this attribute, the first size input element will be used.
+   */
+  get defaultSizeInput(): HTMLInputElement {
+    return (
+      this.sizeTargets.find((input) => 'defaultSize' in input.dataset) ||
+      this.sizeTargets[0]
+    );
+  }
+
+  /**
+   * The currently active device size input element. Falls back to the default size input.
+   */
+  get activeSizeInput(): HTMLInputElement {
+    return (
+      this.sizeTargets.find((input) => input.checked) || this.defaultSizeInput
+    );
+  }
+
+  /**
+   * The URL of the preview iframe and the new tab button.
+   * This takes into account the currently selected preview mode.
+   */
+  get renderUrl(): URL {
+    const url = new URL(this.renderUrlValue, window.location.href);
+    if (this.hasModeTarget) {
+      url.searchParams.set('mode', this.modeTarget.value);
+    }
+    return url;
+  }
+
+  connect() {
```

Comment:
Not critical but I find having the connect method first (after declarations) is helpful.

This way you can kind of read the class end to end a bit easier.

Reply:
I consider getters to be similar to properties, so I put them before `connect()`. In this sense, the `connect()` method is already the first method.

Reply:
Great, no problems.